src/Uri.php

```diff
@@ -356,6 +356,29 @@ public static function withQueryValue(UriInterface $uri, $key, $value)
         return $uri->withQuery(implode('&', $result));
     }
 
+    /**
+     * Creates a new URI with multiple specific query string values.
+     *
+     * Any existing query string values that exactly match the provided key are
+     * removed and replaced with the given key value pair.
+     *
+     * A value of null will set the query string key without a value, e.g. "key"
+     * instead of "key=value".
+     *
+     * @param UriInterface $uri   URI to use as a base.
+     * @param array        $value Associative array of key and values
+     *
+     * @return UriInterface
+     */
+    public static function withQueryValues(UriInterface $uri, $keyValueArray)
+    {
+        foreach ($keyValueArray as $key => $value) {
+            $uri = self::withQueryValue($uri, $key, $value);
```

Comment:
while that implementation is straight forward, it is not optimal as it explodes and implodes the uri each time. please refactor this and move code from withQueryValue to a private method if needed